### PR TITLE
Update to 0.1.0

### PR DESCRIPTION
**Pull request by Antowkos**
You waited for it for a long time and now update is here.
# What's new
* Rework example app and add a bunch of features
* Remove redundant `protocol` word from all protocols
* Simplify `ReuseType` enum
* Add new **context** properties to cell and section items such as
  * `collectionView`
  * `indexPath` for cell items and `index` for section items
  * `sectionItem` for cell items
* Simplify protocol methods and handlers by removing context variables
* Add missed handlers for cell events
* Rework all update methods for cell and section items such as `insert`, `remove`, `reload`
* Add convenient `append` and `prepend` methods to insert items in the end or at the beginning
* Update docs
* Improve performance for sections updates
* Open some internal functions for easy overriding/updating methods
* Fix issue with changing `let` to `var` when setting any handler for cell items
* Update to swift 4.2 syntax
# TODO
- [ ] Improve work with reusable views the same way

### CODE REVIEW

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+            return ClosureWrapper<ActionHandler>.handler(for: self, key: &AssociatedKeys.didDeselectHandler)
+        }
+        set {
+            ClosureWrapper<ActionHandler>.setHandler(newValue, for: self, key: &AssociatedKeys.didDeselectHandler)
+        }
+    }
+    
+    var itemDidDeselectHandler: ActionHandler? {
+        get {
+            return ClosureWrapper<ActionHandler>.handler(for: self, key: &AssociatedKeys.didSelectHandler)
```
Comment (nzatsepilov) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift]: Used wrong associated key. Same in setter.

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            return ClosureWrapper<ActionHandler>.handler(for: self, key: &AssociatedKeys.didUnhighlightHandler)
+        }
+        set {
+            ClosureWrapper<ActionHandler>.setHandler(newValue, for: self, key: &AssociatedKeys.didUnhighlightHandler)
+        }
+    }
+    
+    var itemDidSelectHandler: ActionHandler? {
+        get {
+            return ClosureWrapper<ActionHandler>.handler(for: self, key: &AssociatedKeys.didDeselectHandler)
```
Comment (nzatsepilov) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift]: Used wrong associated key. Same in setter.

**Review by nzatsepilov — CHANGES_REQUESTED** [review]

**Review by nzatsepilov — APPROVED** [review]

File: CollectionViewTools.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -475,13 +478,12 @@
 				DYLIB_INSTALL_NAME_BASE = "@rpath";
 				INFOPLIST_FILE = "$(SRCROOT)/Sources/Info.plist";
 				INSTALL_PATH = "$(LOCAL_LIBRARY_DIR)/Frameworks";
-				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
+				IPHONEOS_DEPLOYMENT_TARGET = 8.2;
```
Comment (evgenmikhaylov) [inline CollectionViewTools.xcodeproj/project.pbxproj:493]: Why 8.2?

File: Example/CollectionViewToolsExample/CellItems/ImageCellItem.swift [hunk 4]
```diff
@@ -5,36 +5,45 @@
...
     
-    func size(for collectionView: UICollectionView, at indexPath: IndexPath) -> CGSize {
-        return collectionView.bounds.size
-    }
-    
-    func cell(for collectionView: UICollectionView, at indexPath: IndexPath) -> UICollectionViewCell {
-        let cell: ImageCollectionViewCell = collectionView.dequeueReusableCell(for: indexPath)
-        cell.imageView.image = image
-        return cell
+    deinit {
```
Comment (evgenmikhaylov) [inline Example/CollectionViewToolsExample/CellItems/ImageCellItem.swift:31]: Does `deinit` need for example? There are too much logs in result

File: Example/CollectionViewToolsExample/CellItems/TextCellItem.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    
+    private static let sizeCell: Cell = .init()
+    func size() -> CGSize {
+        let cell: Cell = type(of: self).sizeCell
+        configure(cell)
+        let cellSize = cell.sizeThatFits(.init(width: collectionView.bounds.size.width, height: .greatestFiniteMagnitude))
+        return .init(width: cellSize.width + 2 * 12, height: collectionView.bounds.height / 1.4)
+    }
+}
```
Comment (evgenmikhaylov) [inline Example/CollectionViewToolsExample/CellItems/TextCellItem.swift:37]: I suppose you should move all variables and typealiases to top and all functions to bottom according to Rosberry code style

**Review by evgenmikhaylov — CHANGES_REQUESTED** [review]

File: Example/CollectionViewToolsExample/Extensions/Collection+Shuffle.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+extension Sequence {
+    /// Returns an array with the contents of this sequence, shuffled.
+    func shuffled() -> [Element] {
+        var result = Array(self)
+        result.shuffle()
+        return result
+    }
+}
```
Comment (artemnovichkov) [inline Example/CollectionViewToolsExample/Extensions/Collection+Shuffle.swift]: Swift 4.2 has these methods out of the box

File: Example/CollectionViewToolsExample/Extensions/Int+Random.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        if range.lowerBound < 0 {
+            offset = abs(range.upperBound)
+        }
+        
+        let minValue = UInt32(range.lowerBound + offset)
+        let maxValue = UInt32(range.upperBound + offset)
+        
+        return Int(minValue + arc4random_uniform(maxValue - minValue)) - offset
+    }
+}
```
Comment (artemnovichkov) [inline Example/CollectionViewToolsExample/Extensions/Int+Random.swift]: Swift 4.2 has these methods out of the box

File: Example/CollectionViewToolsExample/MainViewController.swift [hunk 8]
```diff
@@ -9,51 +9,389 @@ import CollectionViewTools
...
+    var initialImages: [UIImage] {
+        return [#imageLiteral(resourceName: "nightlife-1"), #imageLiteral(resourceName: "nightlife-2"), #imageLiteral(resourceName: "nightlife-3"), #imageLiteral(resourceName: "nightlife-4"), #imageLiteral(resourceName: "nightlife-5")]
+    }
+    var images: [UIImage] {
+        var images: [UIImage] = []
+        for _ in 0..<1 {
+            images.append(contentsOf: shuffledImages)
+        }
+        return images
+    }
```
Comment (artemnovichkov) [inline Example/CollectionViewToolsExample/MainViewController.swift:23]: Why did you add this property?

File: Example/CollectionViewToolsExample/SectionItems/ExampleSectionItem.swift [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+// Copyright (c) 2018 Rosberry. All rights reserved.
+//
+
+import CollectionViewTools
+
+final class ExampleSectionItem: GeneralCollectionViewSectionItem {
+    deinit {
+        print("\(self) deinit")
+    }
+}
```
Comment (artemnovichkov) [inline Example/CollectionViewToolsExample/SectionItems/ExampleSectionItem.swift:11]: Why did you add this class?

File: CollectionViewTools.xcodeproj/project.pbxproj (hunk 3 above)
Comment (antowkos) [reply CollectionViewTools.xcodeproj/project.pbxproj:493]: Because I need the features from 8.2 in example and 8.0 (and even 8.2) isn't supported officially

File: Example/CollectionViewToolsExample/CellItems/ImageCellItem.swift (hunk 4 above)
Comment (antowkos) [reply Example/CollectionViewToolsExample/CellItems/ImageCellItem.swift:31]: I suppose yes, because if it stops logging it means that we made something wrong. But if everyone decide that this is redundant I will remove it.

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: Example/CollectionViewToolsExample/MainViewController.swift (hunk 8 above)
Comment (antowkos) [reply Example/CollectionViewToolsExample/MainViewController.swift:23]: I use it to vary count of generated cells. Should I remove it?

File: Example/CollectionViewToolsExample/SectionItems/ExampleSectionItem.swift (hunk 9 above)
Comment (antowkos) [reply Example/CollectionViewToolsExample/SectionItems/ExampleSectionItem.swift:11]: Just for `deinit` checks

File: Sources/CollectionViewManager.swift [hunk 10]
```diff
@@ -146,69 +126,146 @@ open class CollectionViewManager: NSObject {
...
-        sectionItem.reusableViewItems.forEach { $0.register(for: collectionView) }
+    /// Use this function to force cells and reusable views registration process if you override add/replace/reload methods
+    /// Also in this method section item got set collection view and perform setting index paths and section item for cell items.
+    /// So pay attention to the order of operations for section items updates methods. Set correct index for section item first.
+    /// And don't forget to set `sectionItem` property for cell items.
+    ///
+    /// - Parameter sectionItem: The section item with cell items which need to be registered
+    open func register(_ sectionItem: CollectionViewSectionItem) {
+        sectionItem.collectionView = collectionView
+        sectionItem.cellItems.enumerated().forEach { (index, cellItem) in
```
Comment (rosberry) [inline Sources/CollectionViewManager.swift]: `enumerated` sucks hard in terms of performance. Consider replacing it? 
Not a blocker in any way, but it'll be nice to get rid of it in the future.
Same goes for other occurrences.

File: Sources/CollectionViewManager.swift [hunk 11]
```diff
@@ -217,111 +274,215 @@ extension CollectionViewManager {
     ///   - sectionItem: Section item within which cell items should be replaced
     ///   - completion: A closure that either specifies any additional actions which should be performed after replacing.
     open func replace(cellItemsAt indexes: [Int], with cellItems: [CellItem], in sectionItem: SectionItem, completion: Completion? = nil) {
```
Comment (rosberry) [inline Sources/CollectionViewManager.swift:315]: It'd be great to have `replace` methods which will work with original cellItem instead of indixes,
e.g.: `replace(_ cellItems: [CellItem], with cellItems: [CellItem], ...)`

File: Example/CollectionViewToolsExample/MainViewController.swift (hunk 8 above)
Comment (artemnovichkov) [reply Example/CollectionViewToolsExample/MainViewController.swift:23]: But the count is the same every time. Why didn't you use just `shuffledImages`?

**Comment by artemnovichkov** [comment]
Please, update Swift badge in README

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    var sectionItem: CollectionViewSectionItem? { get set }
+}
+
+extension CollectionViewSiblingCellItem {
+    public var collectionView: UICollectionView {
+        get {
+            if let object = objc_getAssociatedObject(self, &AssociatedKeys.collectionView) as? UICollectionView {
+                return object
+            }
+            fatalError("You should never get this error if you use collection view tools properly. " +
```
Comment (rosberry) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift]: I'm not ok with this. If it's not an optional variable, then it should be safe to access it without considering related implementation details.
Same for other occurrences here.

**Review by artemnovichkov — APPROVED** [review]

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+// MARK: - CollectionViewGeneralCellItem
+
+public typealias ActionHandler = () -> Void
+public typealias ActionResolver = () -> Bool
+public typealias CellActionHandler = (UICollectionViewCell) -> Void
+public typealias ViewActionHandler = (UICollectionReusableView, String, UICollectionView, IndexPath) -> Void
+
+public protocol CollectionViewGeneralCellItem: AnyObject {
```
Comment (rosberry) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift:95]: Can you remind me why we declare it as protocol at all? 
It conforms to `AnyObject` which basically means cell item could only be a class anyway, it requires a ton of associated properties and uses default implementations in extension, which are quite finicky when it comes to subclassing (https://bugs.swift.org/browse/SR-103).

File: Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift [hunk 14]
```diff
@@ -8,15 +8,15 @@ import UIKit.UICollectionView
...
-    open func collectionView(_ collectionView: UICollectionView,
-                             numberOfItemsInSection section: Int) -> Int {
+    open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         return _sectionItems[section].cellItems.count
     }
     
-    open func collectionView(_ collectionView: UICollectionView,
-                             cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    open func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
         let cellItem = self.cellItem(for: indexPath)!
```
Comment (rosberry) [inline Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift:16]: I know it's been here for a long time, but I think it's time to get rid of this force-cast.

File: Sources/CollectionViewManager.swift [hunk 15]
```diff
@@ -217,111 +274,215 @@ extension CollectionViewManager {
...
+            let indexPaths = cellItems.map { cellItem in
+                return cellItem.indexPath
             }
-            return IndexPath(item: row, section: sectionIndex)
-        }
-        
-        collectionView.performBatchUpdates({
-            self.collectionView.deleteItems(at: indexPaths)
-            indexPaths.forEach { sectionItem.cellItems.remove(at: $0.row) }
+            indexPaths.sorted().reversed().forEach { indexPath in
```
Comment (rosberry) [inline Sources/CollectionViewManager.swift]: `sorted(by: >)`?

**Review by ghost — CHANGES_REQUESTED** [review]

File: Sources/CollectionViewManager.swift (hunk 10 above)
Comment (antowkos) [reply Sources/CollectionViewManager.swift]: Are there any proofs of performance lack? I don't mind to replace it but I thought there are no big differences in performance.

File: Sources/CollectionViewManager.swift (hunk 11 above)
Comment (antowkos) [reply Sources/CollectionViewManager.swift:315]: Let's add this as feature request? There are too much of possible cases and convenient improvements for a quick fix.

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift (hunk 13 above)
Comment (antowkos) [reply Sources/Protocols/CellItem/CollectionViewCellItem.swift:95]: All of these protocols weren't conform to `AnyObject` in the beginning. It caused "change `let` to `var`" behaviour after setting any handler and there are only one solution to it - add `AnyObject` conformance.
Anyway I believe that protocols make cell items more flexible. There are no needs to inherit from some class and any object can become `CellItem`.
@artemnovichkov, @Clanat, @medvedzzz let's discuss it and decide what to do?

File: Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift (hunk 14 above)
Comment (antowkos) [reply Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift:16]: Actually I have no idea how to rid it of in elegant way. The only possible solution I see is add `fatalError` if there are no corresponding cell item. 🤔

File: Sources/CollectionViewManager.swift (hunk 15 above)
Comment (antowkos) [reply Sources/CollectionViewManager.swift]: The reason of this operations is to perform removing section items from the end of the array to preserve order of index paths and avoid crashes and unstable behaviour.
Does `sorted(by: >)` do the same?

**Review by evgenmikhaylov — APPROVED** [review]

File: Sources/CollectionViewManager.swift (hunk 10 above)
Comment (disabled) [reply Sources/CollectionViewManager.swift]: It's around 20-25% slower compared to `for in` with a separate var for index. @otbivnoe found it the hard way on pixelcraft.

File: Sources/CollectionViewManager.swift (hunk 15 above)
Comment (disabled) [reply Sources/CollectionViewManager.swift]: `sorted(by: >)` will do the same as `.sorted().reversed()` but shorter and faster :)

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+            objc_setAssociatedObject(self, &AssociatedKeys.indexPath, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
+    }
+    
+    public weak var sectionItem: CollectionViewSectionItem? {
+        get {
+            if let object = objc_getAssociatedObject(self, &AssociatedKeys.sectionItem) as? CollectionViewSectionItem {
+                return object
+            }
+            printContextWarning("We found out that sectionItem property for \(self) is nil")
```
Comment (Otbivnoe) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift:79]: I think it's better to refer to the property via the `#function` keyword, in this case, you don't mind about renaming - all the error messages will be actual.

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+            }
+            printContextWarning("We found out that indexPath property for \(self) is nil")
+            return nil
+        }
+        set {
+            objc_setAssociatedObject(self, &AssociatedKeys.indexPath, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
+    }
+    
+    public weak var sectionItem: CollectionViewSectionItem? {
```
Comment (Otbivnoe) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift:74]: Do you write the `weak` explicitly or not?

```public var collectionView: UICollectionView?```

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+    func didEndDisplaying(cell: UICollectionViewCell) {
+        itemDidEndDisplayingCellHandler?(cell)
+    }
+    
+    func canMove() -> Bool {
+        return itemCanMoveResolver?() ?? false
+    }
+    
+    func size() -> CGSize {
+        return CGSize(width: 50, height: 50)
```
Comment (Otbivnoe) [inline Sources/Protocols/CellItem/CollectionViewCellItem.swift:309]: Why 50x50? Maybe don't we need the default implementation?

File: Sources/Items/GeneralCollectionViewSectionItem.swift [hunk 19]
```diff
@@ -0,0 +1,22 @@
+//
+//  GeneralCollectionViewSectionItem.swift
+//
+//  Copyright © 2017 Rosberry. All rights reserved.
+//
+
+import UIKit.UICollectionView
+
+open class GeneralCollectionViewSectionItem: CollectionViewSectionItem {
```
Comment (Otbivnoe) [inline Sources/Items/GeneralCollectionViewSectionItem.swift:9]: It's better to rename this class to `CollectionViewSectionItem` and add `...Protocol` to the protocol  IMHO. 🤷‍♂️

`General...` 😬

File: Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift [hunk 20]
```diff
@@ -26,14 +26,14 @@ extension CollectionViewManager: UICollectionViewDataSource {
     open func collectionView(_ collectionView: UICollectionView,
                              viewForSupplementaryElementOfKind kind: String,
                              at indexPath: IndexPath) -> UICollectionReusableView {
+        //TODO: Add same `configure` logic as with cell items
```
Comment (Otbivnoe) [inline Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift:29]: With Swift 4.2 looks better to rewrite this TODO to `#warning("TODO: Add same `configure` logic as with cell items")`, there are fewer chances to forget about it.

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift (hunk 18 above)
Comment (antowkos) [reply Sources/Protocols/CellItem/CollectionViewCellItem.swift:309]: It is the default implementation 😄

File: Sources/Items/GeneralCollectionViewSectionItem.swift (hunk 19 above)
Comment (antowkos) [reply Sources/Items/GeneralCollectionViewSectionItem.swift:9]: I don't think so. I did rename all protocols on purpose and roll it back just because of the simplest implementation of `CollectionViewSectionItem` doesn't make sense for me. Adding word `Protocol` in the end isn't an option, but I agree that the name of `GeneralCollectionViewSectionItem` may be changed.

File: Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift (hunk 20 above)
Comment (antowkos) [reply Sources/Extensions/CollectionViewManager+UICollectionViewDataSource.swift:29]: It shouldn't become warning just because of lack of Xcode features 🤔

File: Sources/Protocols/CellItem/CollectionViewCellItem.swift (hunk 16 above)
Comment (antowkos) [reply Sources/Protocols/CellItem/CollectionViewCellItem.swift:79]: Unfortunately it isn't work. `#function` keyword doesn't give us unique and stable pointer 🤷‍♂️

**Review by ghost — APPROVED** [review]
